### PR TITLE
 [Weather] Add two new Provider APIXU and Weatherbit. 

### PR DESCRIPTION
**Pull request by jmahmens**
I 've added two new free provider for weather forecast.

-  APIXU [ApiXU](https://www.apixu.com/)

- Weatherbit [Weatherbit](https://www.weatherbit.io/)

Both provides a free forecast. Only ApiKey must be created like all other provider.
I will try to create openhab2-addons for these both in the next day's.

### CODE REVIEW

**Comment by 9037568** [comment]
First, @jmahmens, thank you for your contribution. However, I'm not sure it makes sense to add new functionality to the OH1 platform, especially when you plan to create new matching OH2 addons.

@kaikreuzer what do you think?

**Comment by jmahmens** [comment]
> 
> First, @jmahmens, thank you for your contribution. However, I'm not sure it makes sense to add new functionality to the OH1 platform, especially when you plan to create new matching OH2 addons.
> 
> @kaikreuzer what do you think?

Well, i think this is a good startingpoint for development of new oh2 addons.
I do this additon of new providers because yahoo was dropped and Wunderground dosn't support free use anymore. OpenWeatherMap provides no free 'daily' forecast, only 3hourly. DarkSky (forecastio), WorldWeaterOnline provides only a limited free forecast.
So this both APIXU and Weatherbit are a chance for those OH1 user have only change their provider and not the whole System.

**Comment by jmahmens** [comment]
Sorry closed accidently

**Comment by kaikreuzer** [comment]
> @kaikreuzer what do you think?

As @jmahmens did the work, I think it is fair to merge it. But in general, I very much agree that new features should rather be done directly as OH2 bindings.

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/common/WeatherConfig.java [hunk 1]
```diff
@@ -93,26 +103,32 @@ public void parse(Dictionary<String, ?> properties) throws ConfigurationExceptio
         for (LocationConfig lc : locationConfigs.values()) {
             if (!lc.isValid()) {
                 parseCompleted = true;
-                logger.warn("Incomplete location config for locationId '{}'. Check openhab.cfg.", lc.getLocationId());
+                logger.warn("Incomplete location config for locationId '{}'. Check openhab.cfg.",
+                    lc.getLocationId());
                 throw new ConfigurationException("weather",
-                        "Incomplete location config for locationId '" + lc.getLocationId() + "'. Check openhab.cfg.");
+                    "Incomplete location config for locationId '" +
+                    lc.getLocationId() + "'. Check openhab.cfg.");
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/common/WeatherConfig.java]: Please change all occurrences of `Check openhab.cfg` to something like `Check the binding config.`

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/converter/property/UnixDateConverter.java [hunk 2]
```diff
@@ -8,26 +8,27 @@
  */
 package org.openhab.binding.weather.internal.converter.property;
 
-import java.util.Calendar;
-
 import org.openhab.binding.weather.internal.converter.Converter;
 import org.openhab.binding.weather.internal.converter.ConverterType;
 
+import java.util.Calendar;
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/converter/property/UnixDateConverter.java]: Revert this change.

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/converter/property/UnixDateConverter.java [hunk 3]
```diff
@@ -8,26 +8,27 @@
...
 public class UnixDateConverter implements Converter<Calendar> {
-
     /**
      * {@inheritDoc}
      */
     @Override
     public Calendar convert(String value) {
         Calendar cal = Calendar.getInstance();
-        cal.setTimeInMillis(Long.valueOf(value) * 1000);
+        cal.setTimeInMillis((long) (Double.valueOf(value) * 1000));
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/converter/property/UnixDateConverter.java]: What's the reason for this?

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/model/Condition.java [hunk 4]
```diff
@@ -8,63 +8,68 @@
...
-import java.util.Calendar;
-
 import org.apache.commons.lang.builder.ToStringBuilder;
 import org.apache.commons.lang.builder.ToStringStyle;
+
 import org.openhab.binding.weather.internal.annotation.Provider;
 import org.openhab.binding.weather.internal.annotation.ProviderMappings;
 import org.openhab.binding.weather.internal.converter.ConverterType;
 
+import java.util.Calendar;
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/model/Condition.java]: Revert this change.

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/provider/AbstractWeatherProvider.java [hunk 5]
```diff
@@ -19,10 +18,15 @@
 import org.openhab.binding.weather.internal.model.ProviderName;
 import org.openhab.binding.weather.internal.model.Weather;
 import org.openhab.binding.weather.internal.parser.WeatherParser;
+
 import org.openhab.io.net.http.HttpUtil;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.Calendar;
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/provider/AbstractWeatherProvider.java]: Revert this change.

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/provider/AbstractWeatherProvider.java [hunk 6]
```diff
@@ -68,45 +79,73 @@ public Weather getWeather(LocationConfig locationConfig) throws Exception {
...
+                locationConfig.getLocationId(), url);
+
+            String response = StringUtils.trimToEmpty(HttpUtil.executeUrl(
+                        "GET", url, 15000));
+
+            /**
+             * special handling because of identical current and forecast json structure
+             * if 'url' contains "forecast" replace data key with forecast
+             */
+            if ((weather.getProvider() == ProviderName.WEATHERBIT) &&
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/provider/AbstractWeatherProvider.java]: Remove the extraneous parentheses.

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/provider/WeatherProviderFactory.java [hunk 7]
```diff
@@ -8,10 +8,11 @@
  */
 package org.openhab.binding.weather.internal.provider;
 
+import org.openhab.binding.weather.internal.model.ProviderName;
```
Comment (9037568) [inline bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/provider/WeatherProviderFactory.java]: Revert this change.

File: bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/converter/property/UnixDateConverter.java (hunk 3 above)
Comment (jmahmens) [reply bundles/binding/org.openhab.binding.weather/src/main/java/org/openhab/binding/weather/internal/converter/property/UnixDateConverter.java]: Weatherbit provides a unix timstamp with fractional miliseconds
like "1547066713.2043"

**Comment by 9037568** [comment]
Thanks, @jmahmens !

**Comment by jmahmens** [comment]
Thanks for mergin this. I will now try the next step. But one question.
Should I try develop an OH2 binding or an eclipse.
A good template for these two new provider are die existing Wunderground and openweathermap eclipse bindings.

**Comment by kaikreuzer** [comment]
@jmahmens Please go for adding them to the openhab2-addons repo - thanks!
